Source/WebCore/Modules/permissions/PermissionStatus.cpp

```diff
@@ -30,49 +30,96 @@
 #include "Document.h"
 #include "DocumentInlines.h"
 #include "EventNames.h"
+#include "MainThreadPermissionObserver.h"
 #include "PermissionController.h"
+#include "PermissionState.h"
+#include "Permissions.h"
+#include "ScriptExecutionContext.h"
 #include "SecurityOrigin.h"
+#include "WorkerGlobalScope.h"
+#include "WorkerLoaderProxy.h"
+#include "WorkerThread.h"
+#include <memory>
+#include <wtf/HashMap.h>
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/MainThread.h>
+#include <wtf/NeverDestroyed.h>
+
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(PermissionStatus);
 
-Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+static HashMap<MainThreadPermissionObserverIdentifier, std::unique_ptr<MainThreadPermissionObserver>>& allMainThreadPermissionObservers()
+{
+    static MainThreadNeverDestroyed<HashMap<MainThreadPermissionObserverIdentifier, std::unique_ptr<MainThreadPermissionObserver>>> map;
+    return map;
+}
+
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page> page)
 {
-    auto status = adoptRef(*new PermissionStatus(context, state, descriptor));
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
+    if (source != PermissionQuerySource::Window)
+        status->createAssociatedMainThreadPermissionObserver(context);
     status->suspendIfNeeded();
     return status;
 }
 
-PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page> page)
     : ActiveDOMObject(&context)
     , m_state(state)
     , m_descriptor(descriptor)
+    , m_source(source)
+    , m_page(page)
 {
     auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
     m_origin = ClientOrigin { context.topOrigin().data(), WTFMove(originData) };
     
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
+    if (source == PermissionQuerySource::Window)
         PermissionController::shared().addObserver(*this);
 }
 
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, weakThis = WeakPtr { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() mutable {
+        if (!weakThis)
```

Comment:
This check is kinda meaningless. `this` gets destroyed on a worker thread so checking if weakThis is null or not on the main thread is super racy.